lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2018, Christian Kotte <christian.kotte@gmx.de>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = r'''
+---
+module: vmware_host_hyperthreading
+short_description: Enables/Disables Hyperthreading optimization for an ESXI host system
+description:
+- This module can be used to enable or disable Hyperthreading optimization for ESXi host systems in given vCenter infrastructure.
+version_added: 2.8
+author:
+- Christian Kotte (@ckotte) <christian.kotte@gmx.de>
+notes:
+- Tested on vSphere 6.5
+requirements:
+- python >= 2.6
+- PyVmomi
+options:
+  enabled:
+     description:
+        - Enable or disable Hyperthreading
+        - You need to reboot the ESXi host if you change the configuration
+        - Make sure that Hyperthreading is enabled in the BIOS. Otherwise, it will be enabled, but never activated.
+     type: bool
+     default: 'yes'
+  esxi_hostname:
+    description:
+    - Name of the host system to work with.
+    - This is required parameter if C(cluster_name) is not specified.
+  cluster_name:
+    description:
+    - Name of the cluster from which all host systems will be used.
+    - This is required parameter if C(esxi_hostname) is not specified.
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- name: Enable Hyperthreading for an host system
+  vmware_host_powermgmt_policy:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    esxi_hostname: '{{ esxi_hostname }}'
+    enabled: true
+    validate_certs: no
+  delegate_to: localhost
+
+- name: Disable Hyperthreading for an host system
+  vmware_host_powermgmt_policy:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    esxi_hostname: '{{ esxi_hostname }}'
+    enabled: false
+    validate_certs: no
+  delegate_to: localhost
+
+- name: Disable Hyperthreading for all host systems from cluster
+  vmware_host_powermgmt_policy:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    cluster_name: '{{ cluster_name }}'
+    enabled: false
+    validate_certs: no
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+result:
+    description: metadata about host system's Hyperthreading configuration
+    returned: always
+    type: dict
+    sample: {
+        "esxi01": {
+            "msg": "Hyperthreading is already enabled and active for host 'esxi01'",
```

Comment:
Add additional information about previous and current state of hyperthreading so that user can do stuff depending on that information in next tasks of playbook.

Reply:
@Akasurde Can you give me an example? I think you implemented this already in one of your other modules!? Otherwise, I would do something like this:
```
results['result'][host.name]['before'] = ...
results['result'][host.name]['after'] = ...
```

Reply:
Something similar - https://github.com/ansible/ansible/blob/fb72a5424cab097d28670d00cf107fd515cee3f0/lib/ansible/modules/cloud/vmware/vmware_host_lockdown.py#L114

Reply:
@Akasurde What's the benefit of `current_state`. Looks like it's similar to `desired_state` all the time!?

Then you can access the state like this?:
```
- name: Enable Hyperthreading
  tags: hyperthreading
  vmware_host_hyperthreading:
    hostname: "{{ vcenter_server }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    esxi_hostname: "{{ inventory_hostname }}"
    enabled: true
    validate_certs: false
  register: hyperthreading_state
  delegate_to: "{{ jump_server }}"

- name: Hyperthreading status
  tags: hyperthreading
  debug:
    var: hyperthreading_state['result']['{{ inventory_hostname }}']['current_state']
```

Reply:
@Akasurde Could you please answer my question? I need that information for other modules as well. Thanks.

Reply:
`current_state` is state after performing vmware task and we can use that information in next ansible task in playbook.
```yaml
  register: hyperthreading_state
  delegate_to: "{{ jump_server }}"

- name: Perform some task depending upon current state of hyperthreading
  debug:
    var: "I am here"
  when: hyperthreading_state['result']['{{ inventory_hostname }}']['current_state'] == 'enabled'
```

Reply:
I think in check mode is like dry-run without any action. so Enable hyperthreading in check mode should have current_state and desired_state as enabled.

Reply:
Done. This just doesn't work if another task depends on `current_state` and would throw an error if Hyperthreading is not active.